lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +802,460 @@ def __str__(self):
         return "[go] %s" % self.url
 
 
+class GitRef(object):
+    """A wrapper that knows how to format git refspecs in multiple contexts."""
+    _known_types = ['branch', 'commit', 'tag']  # type: ClassVar[List[str]]
+    _ref_type = None                            # type: str
+    _ref = None                                 # type: str
+
+    @classmethod
+    def known_types(cls):
+        # type: () -> List[str]
+        return cls._known_types
+
+    @property
+    def ref_type(self):
+        # type: () -> str
+        return self._ref_type
+
+    def __init__(self, ref_type, ref):
+        # type: (str, str) -> None
+        if ref_type not in self.known_types():
+            # This is made a TypeError and not an InvalidGitRef because it cannot arise
+            # from invalid user input.
+            raise TypeError('GitRef can only have the types {0}, but was given {1!r}'
+                            .format(self.known_types(), ref_type))
+        self._ref_type = ref_type
+        if not isinstance(ref, six.string_types):
+            # This case can arise from user error when editing package.py files, so we
+            # want to catch this and handle it specially to provide more context.
+            raise InvalidGitRef('git reference was not a string: {0!r}'.format(ref))
+        self._ref = ref
+
+    def __repr__(self):
+        return '{0}(ref_type={1!r}, ref={2!r})'.format(type(self).__name__,
+                                                       self.ref_type, self._ref)
+
+    def __str__(self):
+        return '<{0}>'.format(self.repo_info_for_reference())
+
+    # Matches any 7-40 character hexadecimal string.
+    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
```

Comment:
I added a commit version regex (slightly different but could be modified) so maybe when one of our PRs is merged we can have this be a global over class variable and share! The one I made is here (doesn't account for the shorter version) https://github.com/spack/spack/pull/24639/files#diff-8dbe8c0b0200a382d9363fe2f8a2133594db50baa72353853027cab92a014d47R49

Reply:
So I decided to keep this separate even after rebasing over your changes, because while the result of a commit hash *from git output with `%H`* will always conform to `^[0-9af]{40}` (and I think we should care whether we are receiving that exact output), I think we want to be able to interpret commit hashes *from within a `package.py`* which may have upper or lower case, and not necessarily full 40-character hashes.